core/src/ast/idents.rs

```diff
@@ -1,43 +1,51 @@
 use proc_macro2::Span;
 use quote::{ToTokens, TokenStreamExt};
 use serde::{Deserialize, Serialize};
-use std::{borrow::Borrow, fmt};
+use std::borrow::{Borrow, Cow};
+use std::fmt;
 
 /// An identifier, analogous to `syn::Ident` and `proc_macro2::Ident`.
 #[derive(Hash, Eq, PartialEq, Serialize, Clone, Debug, Ord, PartialOrd)]
-pub struct Ident(String);
+pub struct Ident(Cow<'static, str>);
 
 impl Ident {
-    /// Create a new `Ident`.
-    ///
-    /// # Panics
-    ///
-    /// This method panics if the provided string isn't a valid identifier.
-    pub fn new(string: &str) -> Self {
-        Ident(syn::parse_str::<syn::Ident>(string).unwrap().to_string())
+    /// Validate a string
+    fn validate(string: &str) {
```

Comment:
suggestion (non-blocking): _Ideally_ we have a constructor that returns a `Result` but I don't think that's too important

Reply:
I considered that but there aren't any places that wouldn't immediately unwrap, so I didn't bother. I'm happy to add if you think it's a good idea though

Reply:
Ideally to do this we'd impl `TryFrom`. But `From` already has a blanket impl for `TryFrom`. And if we made a panicking version and didn't impl `From`, we would need `new_static` and `new` for both `&'static str` and `String`. And this is a lot of work for something we expect to never fail.